app/services/puppet_class_importer.rb

```diff
@@ -171,7 +171,28 @@ def db_environments
   end
 
   def actual_environments
-    @proxy_envs ||= (proxy.environments.map(&:to_s) - ignored_environments)
+    (proxy_environments & (current_user_environments + to_be_created_environments)) - ignored_environments
+  end
+
+  def proxy_environments
+    proxy.environments.map(&:to_s)
+  end
+
+  def to_be_created_environments
+    proxy_environments - Environment.where(:name => proxy_environments).pluck(:name)
+  end
+
+  def current_user_taxonomy_ids
```

Comment:
I agree, that would also fix part of http://projects.theforeman.org/issues/16128